### PR TITLE
New TopBar animation

### LINKED ISSUE #17509: Design feedback for chat topbar
# Bug Report

Address the design review feedback from design on the comments highlighted in these Figma links.
There's a cutout that happens when scrolling to trigger the minified chat nav.

## Resources

Resources
Reported at: https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS?node-id=5963:43613&mode=design#567134213
https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS?node-id=6142:39869&mode=design#567179730

Figma specs: https://www.figma.com/file/7KIYbhoqNGAIFonE0w9TDz/Messages-for-Mobile?type=design&node-id=8451-541849&mode=design&t=Ox3uxC14FqmagMOp-0

### PR DESCRIPTION
**Pull request by ibrkhalil**
fixes #17509 

### Summary

Fixes issue described [here](https://github.com/status-im/status-mobile/issues/17509)

status: ready

### CODE REVIEW

**Comment by ibrkhalil** [comment]
Awaiting for design response to open a PR about this.

**Comment by ibrkhalil** [comment]
Also, I noticed bad UX because of a (vibrating effect).
When there's no more messages at the end.
Should we disable scrolling down until there's actual content

https://github.com/status-im/status-mobile/assets/33176106/5909e445-89da-4c98-ae7c-24ba5481e7c0

**Comment by ibrkhalil** [comment]
Got an okay from design [here](https://discord.com/channels/624634427930312714/928625369542713396/1160961848770830417)

**Review by ajayesivan — APPROVED** [review]
Those numbers look like magic to me. If you could add a brief explanation to assist with future changes, that would be nice.

**Review by alwx — APPROVED** [review]

**Comment by churik** [comment]
Hello @ibrkhalil !
I would not say that animation completely matches design now, but from my perspective, it became better.
Attaching video for better understanding.

Android nightly:

https://github.com/status-im/status-mobile/assets/4557972/6c191ae3-f1b5-4a6c-84e9-215a3aafdb4b

Android PR:

https://github.com/status-im/status-mobile/assets/4557972/e6d1f467-0b21-4f71-91b9-792b0b14aa39

Also, there are still some cases when the top bar is cut:

IMAGE 2023-10-13 18:49:12

But at the same time from my POV the amount of edge cases mentioned https://github.com/status-im/status-mobile/issues/17530 and in https://github.com/status-im/status-mobile/issues/17509 is decreased.

So I'd prefer to get another opinion from a design POV here, no objections to merging this PR from my side.

cc @Francesca-G

**Comment by churik** [comment]
also @ibrkhalil would you mind to add an explanation to your changes as requested in [comment](https://github.com/status-im/status-mobile/pull/17582#pullrequestreview-1666962299)?

**Comment by ibrkhalil** [comment]
> Hello @ibrkhalil ! I would not say that animation completely matches design now, but from my perspective, it became better. Attaching video for better understanding.
> 
> Android nightly:
> 
>  FILE.2023-10-13.18.47.40.mp4 
> Android PR:
> 
>  FILE.2023-10-13.18.48.39.mp4 
> Also, there are still some cases when the top bar is cut:
> 
> IMAGE 2023-10-13 18:49:12
> 
> But at the same time from my POV the amount of edge cases mentioned #17530 and in #17509 is decreased.
> 
> So I'd prefer to get another opinion from a design POV here, no objections to merging this PR from my side.
> 
> cc @Francesca-G



> Those numbers look like magic to me. If you could add a brief explanation to assist with future changes, that would be nice.

I don't think it's optimal to add an explanation to animation values, We're just interpolating on the scroll-y values on a specific threshold. And the threshold is specified based on if it will give us the right animation or not. 

Reanimated docs should explain this in a good way.

**Comment by ibrkhalil** [comment]
> Got an okay from design [here](https://discord.com/channels/624634427930312714/928625369542713396/1160961848770830417)

The green light from design was here, And I think the topbar being cut was meant for the case where the user's name was not visible at all (Cut in half and therefore not readable), And I don't think that's reproducible anymore.

**Comment by Francesca-G** [comment]
> So I'd prefer to get another opinion from a design POV here, no objections to merging this PR from my side.
> 
> cc @Francesca-G

@churik The name of the chat should be visible at all times. When there's not enough messages and the big name is still visible, we shouldn't display any name in the top nav but as soon as the bigger name in the chat is hidden we should display the name in the top nav, with a smooth single animation, avoiding weird states where the name in the top bar isn't fully readable, cut off or with low opacity.

Screenshot 2023-10-16 alle 08 55 01

Screenshot 2023-10-16 alle 08 55 35

**Comment by ibrkhalil** [comment]
> > So I'd prefer to get another opinion from a design POV here, no objections to merging this PR from my side.
> > cc @Francesca-G
> 
> @churik The name of the chat should be visible at all times. When there's not enough messages and the big name is still visible, we shouldn't display any name in the top nav but as soon as the bigger name in the chat is hidden we should display the name in the top nav, with a smooth single animation, avoiding weird states where the name in the top bar isn't fully readable, cut off or with low opacity.
> 
> Screenshot 2023-10-16 alle 08 55 01 Screenshot 2023-10-16 alle 08 55 35

Now I know what to do :)
Thank you so much @churik @Francesca-G

**Review by Francesca-G — APPROVED** [review]
@ibrkhalil nice job ✨

**Comment by churik** [comment]
@ibrkhalil 
There is a bug on IOS - when you tap in composer on empty chat (just right after adding someone to contacts), one the top is shown, and the next time there is no top bar anymore. Issue is reproducible again after reopening chat view.
Video:

https://github.com/status-im/status-mobile/assets/4557972/55820710-9891-4bec-9323-1729e8286fec

**Comment by ibrkhalil** [comment]
@churik Should be ready to test again, Thank you.

File: src/status_im2/contexts/chat/messages/list/view.cljs [hunk 1]
```diff
@@ -318,6 +324,15 @@
        :render-fn                         render-fn
        :on-viewable-items-changed         on-viewable-items-changed
        :on-content-size-change            (fn [_ y]
+                                            (if (or
+                                                 (and keyboard-shown?
+                                                      less-than-two-messages?
+                                                      (> 100 (reanimated/get-shared-value scroll-y))
+                                                      (< 560
+                                                         (reanimated/get-shared-value content-height)))
+                                                 (< 135 (reanimated/get-shared-value scroll-y)))
```
Comment (alwx) [inline src/status_im2/contexts/chat/messages/list/view.cljs]: it's all full of magic numbers 🥲
could you at least move them to the to constants above to make them less random?

**Review by alwx — CHANGES_REQUESTED** [review]

File: src/status_im2/contexts/chat/messages/list/view.cljs (hunk 1 above)
Comment (ibrkhalil) [reply src/status_im2/contexts/chat/messages/list/view.cljs]: Done

**Comment by churik** [comment]
@ibrkhalil 

### ISSUE 1:  issue from https://github.com/status-im/status-mobile/pull/17582#issuecomment-1768412217 is still reproducible if there are several messages in chat (IPhone 14) 

https://github.com/status-im/status-mobile/assets/4557972/2bab39d2-72c1-4e15-a98e-9e9cb9dd108b

### ISSUE 2:  you can reach a state of broken topbar when you expand / collapse composer on empty chat (Android, Google Pixel 7)

IMAGE 2023-10-23 17:18:06

Video:

https://github.com/status-im/status-mobile/assets/4557972/ce56306f-b51e-40f5-ae0b-45ba0885dab3

### ISSUE 3: Fast scrolling back and forth leads to an empty topbar (Android, Google Pixel 7)

https://github.com/status-im/status-mobile/assets/4557972/e395afb5-6f46-469d-9b6e-69d6e7060905

**Comment by ibrkhalil** [comment]
I am going for give this more time to make sure I done it right.

**Comment by ibrkhalil** [comment]
Need another code review please as I did a lot of changes.

**Comment by churik** [comment]
I'm not sure how it correlates with https://github.com/status-im/status-mobile/pull/17734

**Comment by ibrkhalil** [comment]
> I'm not sure how it correlates with #17734

We can merge my PR and then when @alwx creates the abstraction he can rebase and take into consideration the changes made here.

**Review by Francesca-G — COMMENTED** [review]
The animation when scrolling a chat full of messages looks good, I think we can improve the animation when typing messages for the first time in a new chat.
>

https://github.com/status-im/status-mobile/assets/122543089/41d6c3a8-179b-4a54-aa80-c27769fd7c7d

>
As you can see there's a delay in the animation that causes a couple of weird states, like these two: 
>
>
- Name overlapping with an invisible top bar
>
Screenshot 2023-10-30 alle 10 10 10

>
- Duplicate name in the chat view
>
>
Screenshot 2023-10-30 alle 10 10 26

**Comment by ibrkhalil** [comment]
> The animation when scrolling a chat full of messages looks good, I think we can improve the animation when typing messages for the first time in a new chat.
> 
> > 
> 
>  RPReplay_Final1698656958.mp4 
> > 
> 
> As you can see there's a delay in the animation that causes a couple of weird states, like these two:
> 
> > 
> 
> * Name overlapping with an invisible top bar
> 
> > 
> 
> Screenshot 2023-10-30 alle 10 10 10
> > 
> 
> * Duplicate name in the chat view
> 
> > 
> 
> Screenshot 2023-10-30 alle 10 10 26

I think on empty chat, I can trigger the blurring asap and animate the name..
investigating

**Comment by ibrkhalil** [comment]
> The animation when scrolling a chat full of messages looks good, I think we can improve the animation when typing messages for the first time in a new chat.
> 
> > 
> 
>  RPReplay_Final1698656958.mp4 
> > 
> 
> As you can see there's a delay in the animation that causes a couple of weird states, like these two:
> 
> > 
> 
> * Name overlapping with an invisible top bar
> 
> > 
> 
> Screenshot 2023-10-30 alle 10 10 10
> > 
> 
> * Duplicate name in the chat view
> 
> > 
> 
> Screenshot 2023-10-30 alle 10 10 26

Ready for another look 😀

**Comment by ibrkhalil** [comment]
> > The animation when scrolling a chat full of messages looks good, I think we can improve the animation when typing messages for the first time in a new chat.
> 
> > 
> 
> > > 
> 
> > 
> 
> >  RPReplay_Final1698656958.mp4 
> 
> > > 
> 
> > 
> 
> > As you can see there's a delay in the animation that causes a couple of weird states, like these two:
> 
> > 
> 
> > > 
> 
> > 
> 
> > * Name overlapping with an invisible top bar
> 
> > 
> 
> > > 
> 
> > 
> 
> > Screenshot 2023-10-30 alle 10 10 10
> 
> > > 
> 
> > 
> 
> > * Duplicate name in the chat view
> 
> > 
> 
> > > 
> 
> > 
> 
> > Screenshot 2023-10-30 alle 10 10 26
> 
> 
> 
> Ready for another look 😀

Please wait on this, As I found a few other edge cases.

**Comment by ibrkhalil** [comment]
Ready for another review, please.
CC: @churik @Francesca-G

**Comment by ibrkhalil** [comment]
> Ready for another review, please.
> 
> CC: @churik @Francesca-G 

Found another edge case 🥹
Going to make this ready tonight though.

**Review by alwx — APPROVED** [review]

**Comment by ibrkhalil** [comment]
@churik @Francesca-G Let's give testing it a shot please :D

**Comment by Francesca-G** [comment]
@ibrkhalil it's getting better but we still have some issues:


https://github.com/status-im/status-mobile/assets/122543089/e6fdf928-6b3d-43aa-9019-e3dc4f30a907
>
>

Screenshot 2023-11-01 alle 10 31 13

**Comment by ibrkhalil** [comment]
> @ibrkhalil it's getting better but we still have some issues:
> 
> 
> 
> 
> 
> https://github.com/status-im/status-mobile/assets/122543089/e6fdf928-6b3d-43aa-9019-e3dc4f30a907
> 
> >
> 
> >
> 
> 
> 
> Screenshot 2023-11-01 alle 10 31 13
> 
> 

Okay, Another edge case of the two names showing at the same time.
What is the issue in the screenshot?

**Comment by Francesca-G** [comment]
> > @ibrkhalil it's getting better but we still have some issues:
> > 
> >   
> >     
> >     
> > 
> >     RPReplay_Final1698830960.mp4
> >     
> >   
> > 
> >   
> > 
> >   
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > Screenshot 2023-11-01 alle 10 31 13
> 
> Okay, Another edge case of the two names showing at the same time. What is the issue in the screenshot?

The top nav bar is supposed to be there, probably isn't very visible but I have the X button of the top nav overlapping with my avatar :)

**Comment by ibrkhalil** [comment]
Hey @Francesca-G
Can you please take another look?
Thank you :)

**Review by Francesca-G — APPROVED** [review]

**Comment by ibrkhalil** [comment]
CC: @status-im/mobile-qa 
Can I get a review please?

**Comment by churik** [comment]
will take it soon, thank you @ibrkhalil

**Comment by churik** [comment]
@ibrkhalil really good work, thank you!
I'll launch tests on the final commit and will move it to "Merge" once I review them

**Comment by churik** [comment]
Launched tests one more time to be sure that it is really random fails, as I could not find some of them in the other PRs

**Comment by pavloburykh** [comment]
@ibrkhalil PR is ready for merge

**Pull request merged**